exporter/opencensusexporter/opencensus.go

```diff
@@ -16,183 +16,241 @@ package opencensusexporter
-	oce, err := createOCAgentExporter(logger, config, opts...)
+func newTraceExporter(ctx context.Context, cfg *Config) (component.TraceExporterOld, error) {
+	oce, err := newOcExporter(ctx, cfg)
 	if err != nil {
 		return nil, err
 	}
-	oexp, err := exporterhelper.NewMetricsExporterOld(
-		config,
-		oce.PushMetricsData,
-		exporterhelper.WithShutdown(oce.Shutdown))
-	if err != nil {
-		return nil, err
+	oce.traceSvcClient = agenttracepb.NewTraceServiceClient(oce.grpcClientConn)
+	oce.tracesClients = make(chan *tracesClientWithCancel, cfg.NumWorkers)
+	// Try to create rpc clients now.
+	for i := 0; i < cfg.NumWorkers; i++ {
+		tc, errTC := oce.createTraceServiceRPC()
+		if errTC != nil {
+			oce.tracesClients <- nil
+		} else {
+			oce.tracesClients <- tc
+		}
 	}
 
-	return oexp, nil
+	return exporterhelper.NewTraceExporterOld(
+		cfg,
+		oce.pushTraceData,
+		exporterhelper.WithShutdown(oce.shutdown))
 }
 
-func (oce *ocAgentExporter) Shutdown(context.Context) error {
-	wg := &sync.WaitGroup{}
-	var errors []error
-	var errorsMu sync.Mutex
-	visitedCnt := 0
-	for currExporter := range oce.exporters {
-		wg.Add(1)
-		go func(exporter *ocagent.Exporter) {
-			defer wg.Done()
-			err := exporter.Stop()
-			if err != nil {
-				errorsMu.Lock()
-				errors = append(errors, err)
-				errorsMu.Unlock()
-			}
-		}(currExporter)
-		visitedCnt++
-		if visitedCnt == cap(oce.exporters) {
-			// Visited and started Stop on all exporters, just wait for the stop to finish.
-			break
+func newMetricsExporter(ctx context.Context, cfg *Config) (component.MetricsExporterOld, error) {
+	oce, err := newOcExporter(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+	oce.metricsSvcClient = agentmetricspb.NewMetricsServiceClient(oce.grpcClientConn)
+	oce.metricsClients = make(chan *metricsClientWithCancel, cfg.NumWorkers)
+	// Try to create rpc clients now.
+	for i := 0; i < cfg.NumWorkers; i++ {
+		mc, errMC := oce.createMetricsServiceRPC()
+		if errMC != nil {
+			oce.metricsClients <- nil
+		} else {
+			oce.metricsClients <- mc
 		}
 	}
 
-	wg.Wait()
-	close(oce.exporters)
-
-	return componenterror.CombineErrors(errors)
+	return exporterhelper.NewMetricsExporterOld(
+		cfg,
+		oce.pushMetricsData,
+		exporterhelper.WithShutdown(oce.shutdown))
 }
 
-func (oce *ocAgentExporter) PushTraceData(ctx context.Context, td consumerdata.TraceData) (int, error) {
-	// Get first available exporter.
-	exporter, ok := <-oce.exporters
+func (oce *ocExporter) pushTraceData(_ context.Context, td consumerdata.TraceData) (int, error) {
+	// Get first available trace Client.
+	tClient, ok := <-oce.tracesClients
 	if !ok {
-		err := &ocExporterError{
-			code: errAlreadyStopped,
-			msg:  "OpenCensus exporter was already stopped.",
+		err := errors.New("failed to push traces, OpenCensus exporter was already stopped")
+		return len(td.Spans), err
+	}
+
+	if tClient == nil {
```

Comment:
When is this possible? Please add a comment.

Reply:
Done.